File: apple/Packages/BroadwaveKit/Sources/BroadwaveKit/Guide.swift
import Foundation

public enum Category: String, Sendable, CaseIterable {
    case sports, news, movies, kids, series, other

    public var label: String {
        switch self {
        case .sports: "Sports"
        case .news: "News"
        case .movies: "Movies"
        case .kids: "Kids"
        case .series: "Shows"
        case .other: "Other"
        }
    }
}

private let sportsWords = regex(#"\b(sports?|football|basketball|baseball|hockey|soccer|golf|tennis|racing|nascar|motorsports?|boxing|mma|ufc|wrestling|olympics?|nfl|nba|mlb|nhl|mls|wnba|ncaa|bowl|playoffs?|pregame|postgame)\b"#)
private let versus = regex(#"\b(vs\.?|at|@)\b"#)

private func regex(_ pattern: String) -> NSRegularExpression {
    // Patterns above are fixed; a mistake is a programmer error, not a listing error.
    guard let re = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else {
        fatalError("bad pattern")
    }
    return re
}

private func matches(_ re: NSRegularExpression, _ s: String) -> Bool {
    re.firstMatch(in: s, range: NSRange(s.startIndex..., in: s)) != nil
}

public extension Airing {
    /// The same rules as the web guide, so both clients tint and filter alike.
    var kind: Category {
        let c = category ?? ""
        if matches(sportsWords, c) || (matches(sportsWords, title) && matches(versus, title)) {
            return .sports
        }
        if c.localizedCaseInsensitiveContains("news") || title.localizedCaseInsensitiveContains("news") {
            return .news
        }
        if c.localizedCaseInsensitiveContains("movie") || c.localizedCaseInsensitiveContains("film") {
            return .movies
        }
        if ["child", "kids", "animat", "family", "educational"].contains(where: { c.localizedCaseInsensitiveContains($0) }) {
            return .kids
        }
        return c.isEmpty ? .other : .series
    }

    /// "Chiefs at Bills" as a matchup, when the listing names one.
    var matchup: (String, String)? {
        let text = (subtitle.map { $0.range(of: #" (at|vs\.?|@) "#, options: [.regularExpression, .caseInsensitive]) != nil } ?? false) ? subtitle! : title
        guard let r = text.range(of: #"\s+(at|vs\.?|@)\s+"#, options: [.regularExpression, .caseInsensitive]) else { return nil }
        let clean: (Substring) -> String = { s in
            let str = String(s)
            if let colon = str.lastIndex(of: ":") {
                return str[str.index(after: colon)...].trimmingCharacters(in: .whitespaces)
            }
            return str.trimmingCharacters(in: .whitespaces)
        }
        return (clean(text[..<r.lowerBound]), clean(text[r.upperBound...]))
    }

    func minutesLeft(at date: Date) -> String {
        let m = max(0, Int(end.timeIntervalSince(date) / 60))
        return m >= 60 ? "\(m / 60)h \(m % 60)m left" : "\(m)m left"
    }
}

/// Listings by channel, sorted, for fast "what's on" lookups.
public struct GuideIndex: Sendable {
    private var byChannel: [Int64: [Airing]] = [:]

    public init(_ airings: [Airing]) {
        for a in airings {
            byChannel[a.channelId, default: []].append(a)
        }
        for key in byChannel.keys {
            byChannel[key]?.sort { $0.start < $1.start }
        }
    }

    public func airings(_ channel: Int64) -> [Airing] {
        byChannel[channel] ?? []
    }

    public func on(_ channel: Int64, at date: Date) -> Airing? {
        airings(channel).first { $0.isOn(at: date) }
    }

    public func next(_ channel: Int64, after date: Date) -> Airing? {
        airings(channel).first { $0.start >= date }
    }

    /// Program art for a recording: the same program when it has a picture, otherwise the closest listing with the same title.
    public func artAiring(for recording: Recording) -> Airing? {
        let list = airings(recording.channelId)
        if let pid = recording.programId, !pid.isEmpty {
            let matched = list.filter { $0.programId == pid && hasArt($0) }
            if let hit = closest(matched, to: recording.startedAt) {
                return hit
            }
        }
        let titled = list.filter { $0.title == recording.title && hasArt($0) }
        return closest(titled, to: recording.startedAt)
    }
}

private func hasArt(_ airing: Airing) -> Bool {
    guard let image = airing.imageUrl else { return false }
    return !image.isEmpty
}

private func closest(_ list: [Airing], to date: Date) -> Airing? {
    list.min { abs($0.start.timeIntervalSince(date)) < abs($1.start.timeIntervalSince(date)) }
}

public extension Channel {
    /// Channel numbers sort as numbers: 14.2 before 14.10.
    static func guideOrder(_ a: Channel, _ b: Channel) -> Bool {
        let x = a.displayNumber.split(separator: ".").map { Int($0) ?? 0 }
        let y = b.displayNumber.split(separator: ".").map { Int($0) ?? 0 }
        for i in 0 ..< max(x.count, y.count) {
            let l = i < x.count ? x[i] : 0
            let r = i < y.count ? y[i] : 0
            if l != r {
                return l < r
            }
        }
        return a.displayName < b.displayName
    }
}

/// Where the guide grid starts: the half hour before the one `now` is in, so the
/// program on the air and the now line are always on screen.
/// `Calendar.date(bySetting:)` searches forward, so it would pick the next half hour.
public func guideOrigin(for now: Date, calendar: Calendar = .current) -> Date {
    var parts = calendar.dateComponents([.era, .year, .month, .day, .hour, .minute], from: now)
    parts.minute = (parts.minute ?? 0) < 30 ? 0 : 30
    let floored = calendar.date(from: parts) ?? now
    return floored.addingTimeInterval(-30 * 60)
}

File: web/src/lib/guide.ts
import type { Airing, Channel, PlannedAiring, Recording } from "../types";

export type Category = "sports" | "news" | "movies" | "kids" | "series" | "other";

const sportsWords =
  /\b(sports?|football|basketball|baseball|hockey|soccer|golf|tennis|racing|nascar|motorsports?|boxing|mma|ufc|wrestling|olympics?|nfl|nba|mlb|nhl|mls|wnba|ncaa|college (football|basketball)|bowl|playoffs?|pregame|postgame|game day)\b/i;
const vs = /\b(vs\.?|at|@)\b/;

export function categoryOf(airing: Pick<Airing, "category" | "title"> | undefined): Category {
  if (!airing) return "other";
  const c = `${airing.category ?? ""}`;
  if (sportsWords.test(c) || (sportsWords.test(airing.title) && vs.test(airing.title))) return "sports";
  if (/news/i.test(c) || /\bnews\b/i.test(airing.title)) return "news";
  if (/movie|film/i.test(c)) return "movies";
  if (/child|kids|animat|family|educational/i.test(c)) return "kids";
  if (c.trim() !== "") return "series";
  return "other";
}

export const categoryLabel: Record<Category, string> = {
  sports: "Sports",
  news: "News",
  movies: "Movies",
  kids: "Kids",
  series: "Shows",
  other: "Other",
};

export type AiringIndex = Map<number, Airing[]>;

export function indexAirings(airings: Airing[]): AiringIndex {
  const map: AiringIndex = new Map();
  for (const a of airings) {
    let list = map.get(a.channelId);
    if (!list) map.set(a.channelId, (list = []));
    list.push(a);
  }
  for (const list of map.values()) list.sort((a, b) => a.start.localeCompare(b.start));
  return map;
}

export function airingAt(index: AiringIndex, channelId: number, at: number): Airing | undefined {
  const list = index.get(channelId);
  if (!list) return undefined;
  for (const a of list) {
    const s = Date.parse(a.start);
    if (s > at) return undefined;
    if (Date.parse(a.end) > at) return a;
  }
  return undefined;
}

export function nextAfter(index: AiringIndex, channelId: number, at: number): Airing | undefined {
  return index.get(channelId)?.find((a) => Date.parse(a.start) >= at);
}

export function progress(airing: Airing | undefined, now: number): number {
  if (!airing) return 0;
  const s = Date.parse(airing.start);
  const e = Date.parse(airing.end);
  return Math.min(1, Math.max(0, (now - s) / Math.max(1, e - s)));
}

/** Keys of airings that will record or are recording, for guide badges. */
export function recordingKeys(planned: PlannedAiring[], recordings: Recording[]): Set<string> {
  const keys = new Set<string>();
  for (const p of planned) if (!p.skipped && !p.conflict) keys.add(`${p.airing.channelId}@${p.airing.start}`);
  for (const r of recordings) if (r.status === "recording") keys.add(`live:${r.channelId}`);
  return keys;
}

export function isRecording(keys: Set<string>, airing: Airing, now: number): "recording" | "scheduled" | null {
  if (keys.has(`live:${airing.channelId}`) && Date.parse(airing.start) <= now && Date.parse(airing.end) > now) return "recording";
  if (keys.has(`${airing.channelId}@${airing.start}`)) return "scheduled";
  return null;
}

export function sortChannels(channels: Channel[]): Channel[] {
  const num = (v: string) => v.split(".").map((p) => Number(p) || 0);
  return [...channels].sort((a, b) => {
    const x = num(a.displayNumber);
    const y = num(b.displayNumber);
    return (x[0] ?? 0) - (y[0] ?? 0) || (x[1] ?? 0) - (y[1] ?? 0) || a.displayName.localeCompare(b.displayName);
  });
}

export function timeLabel(iso: string | number): string {
  return new Date(iso).toLocaleTimeString([], { hour: "numeric", minute: "2-digit" });
}

export function spanLabel(a: Airing): string {
  return `${timeLabel(a.start)} – ${timeLabel(a.end)}`;
}

export function minutesLeft(a: Airing, now: number): string {
  const m = Math.max(0, Math.round((Date.parse(a.end) - now) / 60000));
  if (m >= 60) return `${Math.floor(m / 60)}h ${m % 60}m left`;
  return `${m}m left`;
}

const sourceLine: Record<string, string> = {
  silicondust: "From the tuner guide.",
  "schedules-direct": "From Schedules Direct.",
  xmltv: "From your guide file.",
  playlist: "From the playlist.",
  broadcast: "From the broadcast.",
};

export function guideSourceLine(source?: string): string {
  return source ? sourceLine[source] ?? "" : "";
}

// emptyGuideLabel is what a row says when the time on screen has no program.
// After the last listing, it names the day the data runs out.
export function emptyGuideLabel(airings: Airing[], windowStart: number): string {
  if (airings.length === 0) return "No listings";
  const end = Date.parse(airings[airings.length - 1].end);
  if (windowStart < end - 60_000) return "No listings";
  const day = new Date(end).toLocaleDateString([], { weekday: "long" });
  return `Listings through ${day}`;
}

export function dayLabel(iso: string, now: number): string {
  const d = new Date(iso);
  const today = new Date(now);
  const startOf = (x: Date) => new Date(x.getFullYear(), x.getMonth(), x.getDate()).getTime();
  const diff = Math.round((startOf(d) - startOf(today)) / 86_400_000);
  if (diff === 0) return "Today";
  if (diff === 1) return "Tomorrow";
  return d.toLocaleDateString([], { weekday: "long", month: "short", day: "numeric" });
}
